crates/json-abi/src/param.rs

```diff
@@ -184,6 +180,12 @@ impl Param {
             s.push_str(&self.ty)
         } else {
             crate::utils::signature_raw("", &self.components, s);
+            // checked during deserialization, but might be invalid from a user
```

Comment:
could we check this on type-instantiation so it's impossible for a user to construct an invalid one?

Reply:
How? All struct fields are public

Reply:
if public fields allow the user to create invalid instances, then we should at least consider making them non-public